File: ros2_ws/src/ugv_bringup_py/ugv_bringup_py/camera_node.py
import cv2
import numpy as np
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge


class CameraNode(Node):
    def __init__(self):
        super().__init__("camera_node")
        # declare parameters
        self.pub_name_img = str(
            self.declare_parameter("pub_name_img", "camera/image_raw").value
        )
        self.timer_period = float(
            self.declare_parameter("timer_period", 0.1).value or 0.1
        )
        self.frame_id = str(self.declare_parameter("frame_id", "pt_camera_link").value)
        self.camera_id = self.declare_parameter("camera_id", 0).value
        self.img_resolution = self.declare_parameter("img_resolution", [640, 480]).value
        self.gst_pipeline = str(self.declare_parameter("gst_pipeline", "").value)

        self.cap = None
        self.last_open_error = None
        self.open_camera()
        # ROS2 components
        self.bridge = CvBridge()
        self.pub_img = self.create_publisher(Image, self.pub_name_img, 10)
        self.timer = self.create_timer(self.timer_period, self.timer_callback)

    def open_camera(self):
        if self.cap is not None:
            return
        cap = None
        if self.gst_pipeline:
            cap = cv2.VideoCapture(self.gst_pipeline, cv2.CAP_GSTREAMER)
        else:
            cap_source = self.camera_id
            if isinstance(cap_source, str) and cap_source.startswith("/dev/video"):
                cap = cv2.VideoCapture(cap_source, cv2.CAP_V4L2)
            else:
                try:
                    cap = cv2.VideoCapture(int(cap_source or 0), cv2.CAP_V4L2)
                except (TypeError, ValueError):
                    cap = cv2.VideoCapture(cap_source)

        if not cap or not cap.isOpened():
            error_msg = f"Failed to open camera source: {self.camera_id}"
            if error_msg != self.last_open_error:
                self.get_logger().error(error_msg)
                self.last_open_error = error_msg
            return

        if (
            isinstance(self.img_resolution, (list, tuple))
            and len(self.img_resolution) == 2
        ):
            cap.set(cv2.CAP_PROP_FRAME_WIDTH, int(self.img_resolution[0]))
            cap.set(cv2.CAP_PROP_FRAME_HEIGHT, int(self.img_resolution[1]))

        self.cap = cap
        self.last_open_error = None
        self.get_logger().info(f"Camera opened: {self.camera_id}")

    def timer_callback(self):
        if self.cap is None:
            self.open_camera()
            return
        ret, frame = self.cap.read()
        if not ret:
            return
        frame = np.ascontiguousarray(frame)
        img_msg = self.bridge.cv2_to_imgmsg(frame, encoding="bgr8")
        img_msg.header.stamp = self.get_clock().now().to_msg()
        img_msg.header.frame_id = self.frame_id
        self.pub_img.publish(img_msg)

    def destroy_node(self):
        if self.cap is not None:
            self.cap.release()
            self.cap = None
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    camera_node = CameraNode()
    rclpy.spin(camera_node)
    camera_node.destroy_node()
    rclpy.shutdown()

File: ros2_ws/src/ugv_bringup_cpp/src/camera_node.cpp
#include <chrono>
#include <memory>
#include <string>
#include <vector>

#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>

namespace ugv_bringup_cpp {

class CameraNode : public rclcpp::Node {
public:
  CameraNode() : Node("camera_node") {
    pub_name_img_ = this->declare_parameter<std::string>("pub_name_img", "camera/image_raw");
    timer_period_sec_ = this->declare_parameter<double>("timer_period", 0.1);
    frame_id_ = this->declare_parameter<std::string>("frame_id", "pt_camera_link");
    camera_id_ = this->declare_parameter<int>("camera_id", 0);
    camera_device_ = this->declare_parameter<std::string>("camera_device", "");
    img_resolution_ = this->declare_parameter<std::vector<int64_t>>("img_resolution",
                                                                     std::vector<int64_t>{640, 480});
    gst_pipeline_ = this->declare_parameter<std::string>("gst_pipeline", "");

    if (timer_period_sec_ <= 0.0) {
      timer_period_sec_ = 0.1;
    }

    pub_img_ = this->create_publisher<sensor_msgs::msg::Image>(pub_name_img_, 10);
    OpenCamera();

    timer_ = this->create_wall_timer(
        std::chrono::duration<double>(timer_period_sec_),
        std::bind(&CameraNode::TimerCallback, this));
  }

  ~CameraNode() {
    if (cap_.isOpened()) {
      cap_.release();
    }
  }

private:
  void OpenCamera() {
    if (cap_.isOpened()) {
      return;
    }
    if (!gst_pipeline_.empty()) {
      cap_.open(gst_pipeline_, cv::CAP_GSTREAMER);
    } else if (!camera_device_.empty()) {
      cap_.open(camera_device_, cv::CAP_V4L2);
    } else {
      cap_.open(camera_id_, cv::CAP_V4L2);
    }

    if (!cap_.isOpened()) {
      RCLCPP_ERROR(this->get_logger(), "Failed to open camera source (id=%d device=%s)",
                   camera_id_, camera_device_.c_str());
      return;
    }

    if (img_resolution_.size() == 2) {
      cap_.set(cv::CAP_PROP_FRAME_WIDTH, static_cast<double>(img_resolution_[0]));
      cap_.set(cv::CAP_PROP_FRAME_HEIGHT, static_cast<double>(img_resolution_[1]));
    }
    cap_.set(cv::CAP_PROP_CONVERT_RGB, 1);
    RCLCPP_INFO(this->get_logger(), "Camera opened.");
  }

  void TimerCallback() {
    if (!cap_.isOpened()) {
      OpenCamera();
      return;
    }
    try {
      cv::Mat frame;
      if (!cap_.read(frame)) {
        RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 5000,
                             "Failed to read frame from camera.");
        return;
      }
      if (frame.empty() || frame.rows <= 0 || frame.cols <= 0 || frame.data == nullptr) {
        return;
      }
      if (frame.channels() == 1) {
        cv::cvtColor(frame, frame, cv::COLOR_GRAY2BGR);
      } else if (frame.channels() == 4) {
        cv::cvtColor(frame, frame, cv::COLOR_BGRA2BGR);
      } else if (frame.channels() != 3) {
        RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 5000,
                             "Unexpected frame channels: %d", frame.channels());
        return;
      }
      if (!frame.isContinuous()) {
        frame = frame.clone();
      }
      auto msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", frame).toImageMsg();
      msg->header.stamp = this->now();
      msg->header.frame_id = frame_id_;
      pub_img_->publish(*msg);
    } catch (const cv::Exception &e) {
      RCLCPP_ERROR_THROTTLE(this->get_logger(), *this->get_clock(), 5000,
                            "OpenCV error: %s", e.what());
      if (cap_.isOpened()) {
        cap_.release();
      }
    }
  }

  std::string pub_name_img_;
  double timer_period_sec_{0.1};
  std::string frame_id_;
  int camera_id_{0};
  std::string camera_device_;
  std::vector<int64_t> img_resolution_;
  std::string gst_pipeline_;

  cv::VideoCapture cap_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_img_;
  rclcpp::TimerBase::SharedPtr timer_;
};

}  // namespace ugv_bringup_cpp

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<ugv_bringup_cpp::CameraNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
